### PR TITLE
langchain: load code from code_paths to system path

### LINKED ISSUE #12922: [BUG] langchain code_paths doesn't load dependencies
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- Client: 2.15.1
- Tracking server: 2.15.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Apple M2 Pro MacOS 14.5
- **Python version**: 3.11.7
- **yarn version, if running the dev UI**:

#### **Logging Model**
```
    logged_chain_info = mlflow.langchain.log_model(
        lc_model=chain_path, artifact_path="chain", model_config="config.yml", code_paths=["helper"]
    )
```

#### **Expected Behavior:**
When loading a model using MLflow with the command:

```python
import mlflow
logged_model = 'runs:/7c16e43f063147b59ea190d1232d0bae/chain'
loaded_model = mlflow.langchain.load_model(logged_model)
```

- The expected behavior is that the model would load successfully, allowing you to use it for predictions or further analysis.
- The code within the model, including any custom modules or helper functions, should be correctly imported and executed without errors.

#### **Actual Behavior:**
- Instead of successfully loading the model, the operation fails with an `MlflowException`.
- The specific issue is a `ModuleNotFoundError` caused by a missing module named `helper`. The model code attempts to import the `extract_chat_history` and `extract_user_query_string` functions from `helper/helper.py`, but the `helper` module is not found in the environment or path where the model is being executed.
- As a result, MLflow cannot complete the model loading process, leading to the following error trace:

```python
ModuleNotFoundError: No module named 'helper'
```

This error is then wrapped and raised as an `MlflowException`:

```python
MlflowException: Failed to import code model from /path/to/temp/dir/chain_as_code.py. Error: No module named 'helper'
```

### Summary

The problem arises because the model code relies on a module (`helper`) that is not available in the current environment or code path when MLflow tries to load the model. This results in a failure to import necessary functions, preventing the model from being loaded and used as intended. 

### Tracking information

```shell
System information: Darwin Darwin Kernel Version 23.5.0: Wed May  1 20:14:38 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6020
Python version: 3.11.7
MLflow version: 2.15.1
MLflow module location: project/.venv/lib/python3.11/site-packages/mlflow/__init__.py
Tracking URI: http://localhost:8080/
Registry URI: http://localhost:8080/
MLflow environment variables: 
  MLFLOW_TRACKING_PASSWORD: 
  MLFLOW_TRACKING_URI: http://localhost:8080/
  MLFLOW_TRACKING_USERNAME: 
MLflow dependencies: 
  Flask: 3.0.3
  Jinja2: 3.1.4
  aiohttp: 3.10.2
  alembic: 1.13.2
  boto3: 1.34.158
  botocore: 1.34.158
  docker: 7.1.0
  graphene: 3.3
  gunicorn: 22.0.0
  langchain: 0.2.10
  markdown: 3.6
  matplotlib: 3.9.1.post1
  mlflow-skinny: 2.15.1
  numpy: 1.26.4
  pandas: 2.2.2
  pyarrow: 15.0.2
  pydantic: 2.8.2
  querystring-parser: 1.2.4
  scikit-learn: 1.5.1
  scipy: 1.14.0
  sqlalchemy: 2.0.32
  tiktoken: 0.7.0
```

### Code to reproduce issue

Code Structure 
```
.
├── chain_as_code.py
├── chain_as_code_driver.py
├── config.yml
└── helper
    └── helper.py
```
`chain_as_code.py`
```
# This example demonstrates defining a model directly from code.
# This feature allows for defining model logic within a python script, module, or notebook that is stored
# directly as serialized code, as opposed to object serialization that would otherwise occur when saving
# or logging a model object.
# This script defines the model's logic and specifies which class within the file contains the model code.
# The companion example to this, chain_as_code_driver.py, is the driver code that performs the  logging and
# loading of this model definition.

import os
from operator import itemgetter

from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import PromptTemplate
from langchain_core.runnables import RunnableLambda
from langchain_openai import OpenAI

import mlflow

assert "OPENAI_API_KEY" in os.environ, "Please set the OPENAI_API_KEY environment variable."

from mlflow.models import ModelConfig
from helper.helper import extract_chat_history, extract_user_query_string

base_config = ModelConfig(development_config="./config.yml")


prompt = PromptTemplate(
    template=base_config.get("prompt_template"),
    input_variables=["question"],
)

model = OpenAI(temperature=0.9)

chain = (
    {
        "question": itemgetter("messages") | RunnableLambda(extract_user_query_string),
        "chat_history": itemgetter("messages") | RunnableLambda(extract_chat_history),
    }
    | prompt
    | model
    | StrOutputParser()
)

mlflow.models.set_model(model=chain)

```

`chain_as_code_driver.py`
```
import mlflow

mlflow.set_tracking_uri("http://localhost:8080")
mlflow.set_experiment("langchain-driver")

# Specify the path to the chain notebook
chain_path = "chain_as_code.py"

print(f"Chain path: {chain_path}")

print("Logging model as code using Langchain log model API")
with mlflow.start_run():
    logged_chain_info = mlflow.langchain.log_model(
        lc_model=chain_path, artifact_path="chain", model_config="config.yml", code_paths=["helper"]
    )

print("Loading model using Langchain load model API")
print(f"Model URI: {logged_chain_info.model_uri}")
model = mlflow.langchain.load_model(logged_chain_info.model_uri)
output = model.invoke(input_example)
print(f"Output: {output}")
```

`config.yml`

```
prompt_template: "You are a hello world bot.  Respond with a reply to the user's question that is fun and interesting to the user.  User's question: {question}"
```

`helper/helper.py`
```
# Return the string contents of the most recent message from the user
def extract_user_query_string(chat_messages_array):
    return chat_messages_array[-1]["content"]


# Return the chat history, which is is everything before the last question
def extract_chat_history(chat_messages_array):
    return chat_messages_array[:-1]
```

When I load the model from a separate folder, it throws error 
```
In [1]: import mlflow
   ...: logged_model = 'runs:/7c16e43f063147b59ea190d1232d0bae/chain'
   ...: loaded_model = mlflow.langchain.load_model(logged_model)
```

### Stack trace

```
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
File project/.venv/lib/python3.11/site-packages/mlflow/models/utils.py:1730, in _load_model_code_path(code_path, config)
   1729     with _mock_dbutils(module.__dict__):
-> 1730         spec.loader.exec_module(module)
   1731 except ImportError as e:

File <frozen importlib._bootstrap_external>:940, in exec_module(self, module)

File <frozen importlib._bootstrap>:241, in _call_with_frames_removed(f, *args, **kwds)

File /var/folders/3d/3jr0cjrd5sg8z_r3kpmh24f00000gp/T/tmp7yr72u0c/chain/chain_as_code.py:24
     23 from mlflow.models import ModelConfig
---> 24 from helper.helper import extract_chat_history, extract_user_query_string
     26 base_config = ModelConfig(development_config="./config.yml")

ModuleNotFoundError: No module named 'helper'

The above exception was the direct cause of the following exception:

MlflowException                           Traceback (most recent call last)
Cell In[1], line 5
      2 logged_model = 'runs:/7c16e43f063147b59ea190d1232d0bae/chain'
      4 # Load model as a PyFuncModel.
----> 5 loaded_model = mlflow.langchain.load_model(logged_model)

File project/.venv/lib/python3.11/site-packages/mlflow/utils/docstring_utils.py:423, in docstring_version_compatibility_warning.<locals>.annotated_func.<locals>.version_func(*args, **kwargs)
    417 if (
    418     warn
    419     and installed_version < Version(min_ver)
    420     or installed_version > Version(max_ver)
    421 ):
    422     _do_version_compatibility_warning(notice)
--> 423 return func(*args, **kwargs)

File project/.venv/lib/python3.11/site-packages/mlflow/tracing/provider.py:237, in trace_disabled.<locals>.wrapper(*args, **kwargs)
    235 disable()
    236 try:
--> 237     is_func_called, result = True, f(*args, **kwargs)
    238 finally:
    239     enable()

File project/.venv/lib/python3.11/site-packages/mlflow/langchain/__init__.py:951, in load_model(model_uri, dst_path)
    929 """
    930 Load a LangChain model from a local file or a run.
    931 
   (...)
    948     A LangChain model instance.
    949 """
    950 local_model_path = _download_artifact_from_uri(artifact_uri=model_uri, output_path=dst_path)
--> 951 return _load_model_from_local_fs(local_model_path)

File project/.venv/lib/python3.11/site-packages/mlflow/langchain/__init__.py:910, in _load_model_from_local_fs(local_model_path, model_config_overrides)
    904 code_path = os.path.join(
    905     local_model_path,
    906     os.path.basename(flavor_code_path),
    907 )
    909 try:
--> 910     model = _load_model_code_path(
    911         code_path, {**(model_config or {}), **(model_config_overrides or {})}
    912     )
    913 finally:
    914     # We would like to clean up the dependencies schema which is set to global
    915     # after loading the mode to avoid the schema being used in the next model loading
    916     _clear_dependencies_schemas()

File project/.venv/lib/python3.11/site-packages/mlflow/models/utils.py:1732, in _load_model_code_path(code_path, config)
   1730         spec.loader.exec_module(module)
   1731 except ImportError as e:
-> 1732     raise MlflowException(
   1733         f"Failed to import code model from {code_path}. Error: {e!s}"
   1734     ) from e
   1735 except Exception as e:
   1736     raise MlflowException(
   1737         f"Failed to run user code from {code_path}. "
   1738         f"Error: {e!s}."
   1739         "Review the stack trace for more information."
   1740     ) from e

MlflowException: Failed to import code model from /var/folders/3d/3jr0cjrd5sg8z_r3kpmh24f00000gp/T/tmp7yr72u0c/chain/chain_as_code.py. Error: No module named 'helper'
```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by harshilprajapati96**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harshilprajapati96/mlflow/pull/12923?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12923/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12923
```

</p>
</details>

Signed-off-by: Harshil Prajapati <harshil.prajapati@gmail.com>

### Related Issues/PRs

 closes #12922 

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/langchain/__init__.py [hunk 1]
```diff
@@ -918,7 +918,7 @@ def _load_model_from_local_fs(local_model_path, model_config_overrides=None):
             # We would like to clean up the dependencies schema which is set to global
             # after loading the mode to avoid the schema being used in the next model loading
             _clear_dependencies_schemas()
-
+        _add_code_from_conf_to_system_path(local_model_path, flavor_conf)
```
Comment (WeichenXu123) [inline mlflow/langchain/__init__.py:921]: The testing code added in this PR seems not covering this line change. Could you create a "code_path" module and make the langchain model code import the "code_path" module  ?

**Review by WeichenXu123 — APPROVED** [review]
LGTM!

**Comment by harshilprajapati96** [comment]
@WeichenXu123  I have updated `validate-labeled`, how I can re-trigger workflow to make sure everything is passing?

File: mlflow/langchain/__init__.py (hunk 1 above)
Comment (WeichenXu123) [inline mlflow/langchain/__init__.py:921]: This line should be moved to the place before L914.

**Review by WeichenXu123 — CHANGES_REQUESTED** [review]
one comment.

**Comment by WeichenXu123** [comment]
> @WeichenXu123 I have updated `validate-labeled`, how I can re-trigger workflow to make sure everything is passing?

I can rerun them.

File: mlflow/langchain/__init__.py (hunk 1 above)
Comment (harshilprajapati96) [reply mlflow/langchain/__init__.py]: fixed!

File: mlflow/langchain/__init__.py [hunk 2]
```diff
@@ -911,14 +911,14 @@ def _load_model_from_local_fs(local_model_path, model_config_overrides=None):
...
                 code_path, {**(model_config or {}), **(model_config_overrides or {})}
             )
         finally:
             # We would like to clean up the dependencies schema which is set to global
             # after loading the mode to avoid the schema being used in the next model loading
             _clear_dependencies_schemas()
-
         return model
     else:
         _add_code_from_conf_to_system_path(local_model_path, flavor_conf)
```
Comment (WeichenXu123) [inline mlflow/langchain/__init__.py:924]: nit: this branch also has `_add_code_from_conf_to_system_path`, we can lift this to the front of `if..else`

File: mlflow/langchain/__init__.py [hunk 3]
```diff
@@ -911,14 +911,14 @@ def _load_model_from_local_fs(local_model_path, model_config_overrides=None):
         )
 
         try:
+            _add_code_from_conf_to_system_path(local_model_path, flavor_conf)
             model = _load_model_code_path(
                 code_path, {**(model_config or {}), **(model_config_overrides or {})}
```
Comment (WeichenXu123) [inline mlflow/langchain/__init__.py:916]: nit: the `code_path` variable name in this line is confusing, could you rename it to `model_code_path` ?

File: mlflow/langchain/__init__.py [hunk 4]
```diff
@@ -906,10 +906,10 @@ def _load_model_from_local_fs(local_model_path, model_config_overrides=None):
             MODEL_CODE_PATH, flavor_conf.get(MODEL_CODE_PATH, None)
         )
         code_path = os.path.join(
-            local_model_path,
+            model_code_path,
```
Comment (WeichenXu123) [inline mlflow/langchain/__init__.py:909]: sorry, I meant rename `code_path` variable name in L908, not the variable `local_model_path` in L909

Comment (harshilprajapati96) [reply mlflow/langchain/__init__.py:909]: do you mean like this?
```
model_code_path = os.path.join(
            local_model_path,
            os.path.basename(flavor_code_path),
        )

model = _load_model_code_path(
                model_code_path, {**(model_config or {}), **(model_config_overrides or {})}
            )
```

**Review by WeichenXu123 — APPROVED** [review]
LGTM

File: mlflow/langchain/__init__.py [hunk 5]
```diff
@@ -905,23 +905,21 @@ def _load_model_from_local_fs(local_model_path, model_config_overrides=None):
         flavor_code_path = pyfunc_flavor_conf.get(
             MODEL_CODE_PATH, flavor_conf.get(MODEL_CODE_PATH, None)
         )
-        code_path = os.path.join(
+        model_code_path = os.path.join(
             local_model_path,
             os.path.basename(flavor_code_path),
         )
-
+        _add_code_from_conf_to_system_path(local_model_path, flavor_conf)
```
Comment (WeichenXu123) [inline mlflow/langchain/__init__.py:912]: It should be moved to front of L896 because you removed L924

Comment (WeichenXu123) [reply mlflow/langchain/__init__.py:912]: @harshilprajapati96 Could you also improve test to make it cover this case ? :)

Comment (harshilprajapati96) [reply mlflow/langchain/__init__.py]: added test to cover the else condition!

**Comment by WeichenXu123** [comment]
LGTM

**Review by WeichenXu123 — APPROVED** [review]

**Pull request merged**